### PR TITLE
Problem: contract suicide is not tested

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #285
Solution:
test that:
- within the tx, after the contract suicided, the code still exists.
- after the tx, the code no longer exist.

BTW, this test case don't pass for current ethermint, it's only fixed in statedb-refactor branch.

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

**Review by tomtau — COMMENTED** [review]
probably good to link the reference for how the contract suicide should happen (in the yellow paper?)

**Comment by yihuang** [comment]
> probably good to link the reference for how the contract suicide should happen (in the yellow paper?)

the same `test_suicide` case will run on both geth and cronos (because of the parametrized `cluster` fixture), to check identical behaviors on both chains.

File: integration_tests/test_basic.py [hunk 1]
```diff
@@ -427,3 +428,28 @@ def test_message_call(cronos):
...
+
+
+def test_suicide(cluster):
+    """
+    test compliance of contract suicide
+    - within the tx, after contract suicide, the code is still available.
+    - after the tx, the code is not available.
+    """
+    w3 = cluster.w3
+    inner = deploy_contract(
```
Comment (damoncro) [inline integration_tests/test_basic.py:440]: what inner and outer means?

Comment (yihuang) [reply integration_tests/test_basic.py:440]: they are just the contract names: https://github.com/crypto-org-chain/cronos/pull/286/files#diff-691042267b4f7caaa41185a26ffad9fae48b8960f4c668bba805f4cb782f527fR4
inner contract is to be destroyed by the outer contract.

Comment (tomtau) [reply integration_tests/test_basic.py:440]: I assumed that after reading through it, but was puzzled by the names as well 😂 perhaps a comment there or renaming it as "to_be_destroyed"/"destroyee" and "destroyer" contracts?

Comment (yihuang) [reply integration_tests/test_basic.py]: sounds good, renamed.

**Comment by yihuang** [comment]
> probably good to link the reference for how the contract suicide should happen (in the yellow paper?)

actually, I didn't find the very specification about this on yellow paper.

**Comment by adu-crypto** [comment]
I do find some descriptions related to self-destructs that imply the code of account should be cleared after tx execution:
```
The tuple contents include As, the self-destruct set: a
set of accounts that will be discarded following the transaction’s completion
```

**Comment by adu-crypto** [comment]
https://github.com/ethereum/go-ethereum/blob/443afc975c7a4016bf9ed1b057b8cc186cb089e1/core/state/statedb.go#L433
```
// Suicide marks the given account as suicided.
// This clears the account balance.
//
// The account's state object is still available until the state is committed,
// getStateObject will return a non-nil account after Suicide.
```
so before the commitment, the code size should not be zero as the test suggested.

**Comment by yihuang** [comment]
> I do find some descriptions related to self-destructs that imply the code of account should be cleared after tx execution:
> 
> ```
> The tuple contents include As, the self-destruct set: a
> set of accounts that will be discarded following the transaction’s completion
> ```

Did you find spec about the behavior between the selfdestruct and tx commit?

**Comment by adu-crypto** [comment]
I find this:
Screen Shot 2022-01-04 at 3 49 29 PM

it says:
```
halt execution and register account for later deletion
```

**Comment by adu-crypto** [comment]
go-ethereum just append suicideChange message to the journal and mark stateObject as suicided.
https://github.com/ethereum/go-ethereum/blob/443afc975c7a4016bf9ed1b057b8cc186cb089e1/core/state/statedb.go#L440
I guess any changes to statedb should not take effect until journal commitment?

**Comment by yihuang** [comment]
> go-ethereum just append suicideChange message to the journal and mark stateObject as suicided. https://github.com/ethereum/go-ethereum/blob/443afc975c7a4016bf9ed1b057b8cc186cb089e1/core/state/statedb.go#L440 I guess any changes to statedb should not take effect until journal commitment?

it's marked as suicided first, only be actually deleted when commit.

**Comment by adu-crypto** [comment]
> 

yes

**Review by tomtau — APPROVED** [review]

**Pull request merged**